src/crawlers/browser_crawler.js

```diff
@@ -0,0 +1,523 @@
+            postLaunchHooks: [
+                this._maybeAddSessionRetiredListener.bind(this),
+                ...postLaunchHooks,
+            ],
+        });
+    }
+
+    /**
+     * Wrapper around handlePageFunction that opens and closes pages etc.
+     *
+     * @param {Object} crawlingContext
+     * @param {Request} crawlingContext.request
+     * @param {BrowserController} crawlingContext.browserController
+     * @param {Session} [crawlingContext.session]
+     * @private
+     */
+    async _handleRequestFunction(crawlingContext) {
+        const { id } = crawlingContext;
+        const page = await this.browserPool.newPage({ id });
+        this._enhanceCrawlingContextWithPageInfo(crawlingContext, page);
+
+        const { request, session } = crawlingContext;
+
+        if (this.persistCookiesPerSession) {
+            const cookies = crawlingContext.session.getPuppeteerCookies(request.url);
+            await crawlingContext.browserController.setCookies(page, cookies);
+        }
+
+        try {
+            await this._handleNavigation(crawlingContext);
+
+            await this._responseHandler(crawlingContext);
+
+            // save cookies
+            // @TODO: Should we save the cookies also after/only the handle page?
+            if (this.persistCookiesPerSession) {
+                const cookies = await crawlingContext.browserController.getCookies(page);
+                session.setPuppeteerCookies(cookies, request.loadedUrl);
+            }
+
+            await addTimeoutToPromise(
+                this.handlePageFunction(crawlingContext),
+                this.handlePageTimeoutMillis,
+                `handlePageFunction timed out after ${this.handlePageTimeoutMillis / 1000} seconds.`,
+            );
+
+            if (session) session.markGood();
+        } finally {
+            try {
+                await page.close();
+            } catch (error) {
+                // Only log error in page close.
+                this.log.debug('Error while closing page', { error });
+            }
+        }
+    }
+
+    /**
+     *
+     * @param {object} crawlingContext
+     * @param {Page} page
+     * @private
+     */
+    _enhanceCrawlingContextWithPageInfo(crawlingContext, page) {
+        crawlingContext.page = page;
+
+        // This is the wierd spam because of browser to proxy not page to proxy.
+        const browserControllerInstance = this.browserPool.getBrowserControllerByPage(page);
+        crawlingContext.browserController = browserControllerInstance;
+
+        crawlingContext.session = browserControllerInstance.launchContext.session;
+        crawlingContext.proxyInfo = browserControllerInstance.launchContext.proxyInfo;
+
+        crawlingContext.crawler = this;
+    }
+
+    /**
+     *
+     * @param {object} crawlingContext
+     * @private
+     */
+    async _handleNavigation(crawlingContext) {
+        // @TODO: consider deep clone
+        const gotoOptions = _.clone(this.gotoOptions);
```

Comment:
let's just use deepClone, the opts are tiny

Reply:
The issue is that underscore does not have deep clone 😞

Reply:
Let's do it with the spread.